firedrake/adjoint/function.py

```diff
@@ -121,8 +121,11 @@ def wrapper(self, other, *args, **kwargs):
                 block_var = self.create_block_variable()
                 block.add_output(block_var)
 
-                if isinstance(other, type(self)):
-                    block_var._checkpoint = DelegatedFunctionCheckpoint(other.block_variable)
+                if (isinstance(other, type(self))):
+                    other_mesh = other.function_space().mesh()
+                    self_mesh = self.function_space().mesh()
+                    if other_mesh == self_mesh:
```

Comment:
```suggestion
                    if other.function_space().mesh() == self.function_space().mesh():
```
Don't introduce unnecessary variables.